### PR TITLE
Add heuristics for C++ macros

### PR DESCRIPTION
**Pull request by jtibshirani**
This PR adds a new heuristic for C++ that checks for the preprocessor macros
__has_cpp_attribute and __cplusplus. These are commonly-used macros that check
what version of C++ is supported, and the available attributes.

## Checklist:
- [x] **I am fixing a misclassified language**
  - [x] I have included a new sample for the misclassified language:
    - Sample source(s): [NoDiscard.h](https://github.com/apple/swift/blob/f9ec3b1d7e6e4698b73120a548e93adf22b097e8/include/swift/Basic/NoDiscard.h#L4) from Swift project
    - Sample license(s): Apache v2
  - [x] I have included a change to the heuristics to distinguish my language from others using the same extension.

### CODE REVIEW

File: test/test_blob.rb [hunk 1]
```diff
@@ -260,7 +260,7 @@ def test_vendored
 
   def test_language
     allowed_failures = {
-      "#{samples_path}/C++/rpc.h" => ["C", "C++"],
+      "#{samples_path}/C/rpc.h" => ["C", "C++"],
```
Comment (jtibshirani) [inline test/test_blob.rb:263]: This is not related to my change, I just noticed these paths were out-of-date.

**Review by lildude — APPROVED** [review]
LGTM. Thanks.

Note: this PR will not be merged until close to when the next release is made. See [here](https://github.com/github/linguist/blob/master/docs/troubleshooting.md#my-linguist-pr-has-been-merged-but-github-doesnt-reflect-my-changes) for more details.

**Comment by jtibshirani** [comment]
Thanks for the speedy review!
